sourcegraph/sourcegraph
# feat: generate endpoints using replica numbers

RE:  https://github.com/sourcegraph/deploy-sourcegraph/pull/4213 & https://github.com/sourcegraph/sourcegraph/issues/45861

## Summary

**This feature is also backward-compatible, so existing setups and instances will not be affected by this change.**

This PR adds the ability to accept replicas number for `Gitserver`, `Indexed-search`, `Searcher`, and `Symbols` via existing env vars (`SEARCHER_URL`, `SYMBOLS_URL`, `INDEXED_SEARCH_SERVERS`, `SRC_GIT_SERVERS`), and then generate a new list of endpoints based on replicas number instead of listing out each replica endpoint manually. 

### Background

Issue: Our current default setup only works in RBAC-enabled clusters. 

RBAC for frontend is used for service discovery as it is essential for frontend to arrange communications between services; however, this default setup often leads to delays and frustrating deployment experiences for those without the ability to create RBAC resources. As this is a common deployment pain point and issue that has negative impacts on both the customers and our business, our team believes this is something we should address. 

### Proposed Solution

A new base cluster introduces in https://github.com/sourcegraph/deploy-sourcegraph/pull/4215 provide an option for users to set up a working Sourcegraph k8s instance with or without RBAC enabled for frontend. 

When RBAC resources are removed from frontend, Sourcegraph will not work properly unless a list of services endpoints is included in the frontend env var list using existing env vars that are also being used in our[ docker-compose deployment ](https://github.com/sourcegraph/deploy-sourcegraph-docker/blob/master/docker-compose/docker-compose.yaml#L136-L147), which is a task that relies on users inputting the correct values manually.

The functions introduce in this PR aim to minimize the time our users/customers have to spend configuring their deployment (e.g. spending time inputting the endpoint addresses for each service manually) by generating the endpoints in frontend based on the replica number provided via env vars. 

**IMPORTANT** The changes made in this PR should **NOT** break any existing setup or instances. No additional changes are required in existing manifests.

### Expected Behavior

The current behavior is the same for **all** instances (both existing instances and new instances), which means:
- **ALL**  instances that do not have the **existing** `endpoints URLs` env vars set are default to using Kubernetes service discovery 
- Setting env vars to empty string "" will return ""
- See the Examples section listed below for the expected behaviors

#### Examples:

Return endpoints generated with the replica number:
```yaml
        env:
        - name: SRC_GIT_SERVERS
          value: "4"
        - name: SEARCHER_URL
          value: "4"
        - name: SYMBOLS_URL
          value: "4"
        - name: INDEXED_SEARCH_SERVERS
          value: "4"
```


## Test plan



1. copy and paste the context below to your `sg.config.overwrite.yaml` file in the root of this directory:
2. run `sg start enterprise`
3. You should see frontend and gitserver complaining about the new endpoints that do not exist, with the newly generated list of endpoints in the correct format for k8s deployments.

The expected new endpoints using the file shared below should look like:
1. 4 gitserver endpoints: `"gitserver-0.gitserver:3178", "gitserver-1.gitserver:3178", "gitserver-2.gitserver:3178", "gitserver-3.gitserver:3178", "gitserver-4.gitserver:3178"`
2. 3 indexed search endpoints: `"indexed-search-0.indexed-search:6070", "indexed-search-1.indexed-search:6070", "indexed-search-2.indexed-search:6070"`
3. 3 symbols endpoints: `"http://symbols-0.symbols:3184", "http://symbols-1.symbols:3184", "http://symbols-2.symbols:3184"`
4. 3 searcher endpoints: `"http://searcher-0.searcher:3181", "http://searcher-1.searcher:3181", "http://searcher-2.searcher:3181"`

`sg.config.overwrite.yaml`:
```yaml
env:
  SRC_GIT_SERVERS: "5"
  SEARCHER_URL: "3"
  SYMBOLS_URL:"3"
  INDEXED_SEARCH_SERVERS: "3"
  DEPLOY_TYPE: kubernetes
```

Example output from running sg start:
```sh
[    gitserver-1] WARN server server/cleanup.go:193 current shard is not included in the list of known gitserver shards, will not delete repos {"current-hostname": "127.0.0.1:3502", "all-shards": ["gitserver-0.gitserver:3178", "gitserver-1.gitserver:3178", "gitserver-2.gitserver:3178", "gitserver-3.gitserver:3178", "gitserver-4.gitserver:3178"]}
[    gitserver-0] WARN server server/cleanup.go:193 current shard is not included in the list of known gitserver shards, will not delete repos {"current-hostname": "127.0.0.1:3501", "all-shards": ["gitserver-0.gitserver:3178", "gitserver-1.gitserver:3178", "gitserver-2.gitserver:3178", "gitserver-3.gitserver:3178", "gitserver-4.gitserver:3178"]}
[       frontend] ERROR InternalHandler httpapi/httpapi.go:281 API HTTP handler error response {"method": "POST", "request_uri": "/.internal/repos/index", "status_code": 500, "error": "hostname \"localhost:3070\" not found in Endpoints{\"indexed-search-0.indexed-search:6070\" \"indexed-search-1.indexed-search:6070\" \"indexed-search-2.indexed-search:6070\"}"}
[       frontend] ERROR internal.http trace/httptrace.go:250 unexpected status code 500 {"route_name": "internal.repos.index", "method": "POST", "url": "/.internal/repos/index", "code": 500, "duration": "2.073375ms", "shouldTrace": false, "error": "hostname \"localhost:3070\" not found in Endpoints{\"indexed-search-0.indexed-search:6070\" \"indexed-search-1.indexed-search:6070\" \"indexed-search-2.indexed-search:6070\"}"}
```
